components/org.wso2.carbon.metrics.impl/src/main/java/org/wso2/carbon/metrics/impl/MetricServiceImpl.java

```diff
@@ -378,10 +417,112 @@ private boolean isMetricEnabledBasedOnHierarchyLevel(String name, Level metricLe
     }
 
     /**
+     * Get or create a metric collection for a given path
+     *
+     * @param annotatedName The annotatedName of the metric
+     * @param levels The {@code Level}s for affected metrics
+     * @param metricBuilder A {@code MetricBuilder} instance used to create the relevant metric
+     * @return The created {@link Metric} collection
+     */
+    @SuppressWarnings("unchecked")
+    private <T extends AbstractMetric> Metric getOrCreateMetricCollection(String annotatedName, Level[] levels,
+            MetricBuilder<T> metricBuilder) throws MetricNotFoundException {
+        Level level = null;
+        if (levels != null && !isLevelsMatch(annotatedName, levels)) {
+            throw new IllegalArgumentException("number of metric levels doesn't match the annotated name");
+        } else if (levels != null && levels.length > 0) {
+            level = levels[levels.length - 1];
+        }
+        Metric metricCollection = metricsCollections.get(annotatedName);
+        if (metricCollection == null) {
+            String name = annotatedName.replaceAll("\\[\\+\\]", "");
+            Metric metric;
+            if (level != null) {
+                metric = getOrCreateMetric(name, level, metricBuilder);
+            } else {
+                metric = getMetric(name, metricBuilder);
+            }
+            boolean annotated = isAnnotated(annotatedName);
+            List<?> affected = getAffectedMetrics(annotatedName, levels, metricBuilder);
+            if (annotated && metric instanceof Counter) {
+                metricCollection = new CounterCollection((Counter) metric, (List<Counter>) affected);
+                metricsCollections.put(annotatedName, metricCollection);
+            } else if (annotated && metric instanceof Histogram) {
+                metricCollection = new HistogramCollection((Histogram) metric, (List<Histogram>) affected);
+                metricsCollections.put(annotatedName, metricCollection);
+            } else if (annotated && metric instanceof Meter) {
+                metricCollection = new MeterCollection((Meter) metric, (List<Meter>) affected);
+                metricsCollections.put(annotatedName, metricCollection);
+            } else {
+                metricCollection = metric;
+            }
+        }
+        return metricCollection;
+    }
+
+    private boolean isAnnotated(String annotatedName) {
+        return annotatedName.contains("[+]");
+    }
+
+    private boolean isLevelsMatch(String annotatedName, Level[] levels) {
+        Pattern p = Pattern.compile("\\[\\+\\]");
```

Comment:
Compiling patterns per invocation can be expensive. Instead cache the compiled pattern and re-use.